File: codigo final/Java/src/serie2/ConcurrentQueue.java
package serie2;

import java.util.concurrent.atomic.AtomicReference;

public class ConcurrentQueue<E> {
    private static class Node<T> {
        final T item;
        final AtomicReference<Node<T>> next;
        public Node(T item, Node<T> next) {
            this.item = item;
            this.next = new AtomicReference<>(next);
        }
    }

    private final AtomicReference<Node<E>> head;
    private final AtomicReference<Node<E>> tail;

    public ConcurrentQueue() {
        Node<E> dummy = new Node<>(null, null);
        head = new AtomicReference<>(dummy);
        tail = new AtomicReference<>(dummy);
    }

    public void put(E item) {
        Node<E> newNode = new Node<>(item, null);
        while (true) {
            Node<E> currentTail = tail.get();
            Node<E> nextTail = currentTail.next.get();
            if(tail.get() == currentTail) {
                if(nextTail != null)
                    tail.compareAndSet(currentTail, nextTail);
                else {
                    if(currentTail.next.compareAndSet(null, newNode)) {
                        tail.compareAndSet(currentTail, newNode);
                        return;
                    }
                }
            }
        }
    }

    public E tryTake() {
        while (true) {
            Node<E> currentHead = head.get();
            Node<E> currentTail = tail.get();
            Node<E> nextHead = currentHead.next.get();

            if(currentHead == head.get()) {
                if(currentHead == currentTail) {
                    if(nextHead == null)
                        return null;

                    tail.compareAndSet(currentTail, nextHead);
                } else   {
                    E value = nextHead.item;
                    if(head.compareAndSet(currentHead, nextHead))
                        return value;
                }
            }
        }
    }

    public E take() throws InterruptedException {
        E value;
        while ((value = tryTake()) == null)
            Thread.sleep(0);
        return value;
    }

    public boolean isEmpty() {
        return tail.get() == head.get();
    }
}

File: codigo final/C#/PC/serie2/ConcurrentQueue.cs
using System.Threading;

namespace serie2
{
    public class ConcurrentQueue<T>
    { 
        private class Node<T> {
            public readonly T item;
            public volatile Node<T> next;

            public Node(T item, Node<T> next) {
                this.item = item;
                this.next = next;
            }
        }

        private readonly Node<T> dummy;
        private volatile Node<T> head;
        private volatile Node<T> tail;

        public ConcurrentQueue() {
            dummy = new Node<T>(default(T), null);
            head = dummy;
            tail = dummy;
        }
        
        public void Put(T item) {
            var newNode = new Node<T>(item, null);

            while (true) {
                var currentTail = tail;
                var nextTail = currentTail.next;

                if (currentTail == tail)
                    if (nextTail != null)
                        Interlocked.CompareExchange(ref tail, nextTail, currentTail);
                    else
                        if (Interlocked.CompareExchange(ref currentTail.next, newNode, null) == null) {
                            Interlocked.CompareExchange(ref tail, newNode, currentTail);
                            return;
                        }
            }
        }

        private T TryTake() {
            while (true) {
                var currentHead = head;
                var currentTail = tail;
                var nextHead = currentHead.next;

                if (currentHead == head) {
                    if (currentHead == currentTail) {
                        if (nextHead == null)
                            return default(T);

                        Interlocked.CompareExchange(ref tail, nextHead, currentTail);
                    } else {
                        var value = nextHead.item;

                        if (Interlocked.CompareExchange(ref head, nextHead, currentHead) == currentHead)
                            return value;
                    }
                }
            }
        }

        public T Take() {
            T value;
            while ((value = TryTake()) == null)
                Thread.Sleep(0);
            return value;
        }

        public bool IsEmpty() {
            return head == tail;
        }
    }
}